### DISCUSSION TITLE
Trouble implementing home row mods

### DISCUSSION THREAD

**Discussion by maykot** [houmain/keymapper#187] [Q&A] [answered]
I've been trying to implement home row mods using keymapper and I hit a wall.

My first naive implementation was just something like this: `A{Any} >> Meta{Any}`. Which is no good, it makes typing very difficult because rolling `A Any` (i.e. `press A, press Any, release A, release Any`) causes `Meta{Any}` to be sent. To fix this I've been trying to implement something like [kanata](https://github.com/jtroo/kanata/tree/main)'s `tap-hold-release`, without success. But now I'm not even sure that's a good idea.

Ideally I'd want `A` to act as `Meta` whenever the following sequence of events happens: `press A, press Any, release Any` (with other mods optionally held as well). And importantly, the following sequence should **not** trigger `Meta` and should default to normal behavior: `press A, press Any, release A`.

Has anyone else implemented this already? Any help would be much appreciated.

**Comment by ristomatti** [comment]
I have not implemented Home Row mods specifically, but I have implemented modifier like mappings on a couple of alpha keys. I've been using them for quite some time now with neglible effect on my typing. My full config is on GitHub but the config is quite complex and might not be up-to-date on this part. 

I've tried to extract the relevant bits of my most used one that allows using `HJKL` as arrow keys while holding `E`. I didn't test this by itself though. 

```ini
# Any modifier
Modifier    = Control | Meta | Shift | Alt | AltGr

# Cursor movement
CursorLeft  = H
CursorDown  = J
CursorUp    = K
CursorRight = L

# Virtual mods
EditMode    = Virtual20

# Alpha mods
Edit        = E

# Macros
Tap         = $0{!120ms}
Hold        = $0{130ms}

# Edit mode
[modifier = "EditMode"]
  CursorUp    >> ArrowUp
  CursorLeft  >> ArrowLeft
  CursorDown  >> ArrowDown
  CursorRight >> ArrowRight
  Space       >> Space

[modifier = "!Modifier !Space"]
  ? Tap[Edit] >> Edit
  Hold[Edit]  >> EditMode ^ EditMode
```

Negated `Modifier` is obviously to avoid triggering the mode when `E` is used in combination of standard modifiers.

I don't remember exactly if the stuff involving `Space` was related to my usage of `Space` as the `Meta` key further down the config, or if it caused issues with this particular mapping. It might not be needed.

**Reply by ristomatti** [reply]
I believe the relevant part for your use case is the last context block. Using the `?` operator allows the key to trigger instantly when typing normally. I found this combination of timed key release/hold behavior to help with the mapping getting in the way of typing. You'll likely want to adjust the timings to match your typing speed.

**Reply by ristomatti** [reply]
What you might also want is to define the set of keys you need when pressing `Meta`. Here's a set of aliases that might help making the mappings more concise if you wish to do that:

```starlark
# Key groups
Alpha           = A | B | C | D | E | F | G | H | I | J | K | L | M | N | O | P | Q | R | S | T | U | V | W | X | Y | Z
Digit           = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
Symbol          = Backquote | BracketLeft | BracketRight | Comma | Equal | Minus | Period | Semicolon | Slash | Quote
Character       = Alpha | Digit | Symbol
Whitespace      = Space | Tab | Enter

NumpadDigit     = Numpad0 | Numpad1 | Numpad2 | Numpad3 | Numpad4 | Numpad5 | Numpad6 | Numpad7 | Numpad8 | Numpad9
NumpadSymbol    = NumpadDivide | NumpadMultiply | NumpadSubtract | NumpadAdd | NumpadEnter | NumpadDecimal

ArrowKey        = ArrowUp | ArrowDown | ArrowLeft | ArrowRight
NavigationKey   = ArrowKey | Home | End | PageUp | PageDown
FunctionKey     = F1 | F2 | F3 | F4 | F5 | F6 | F7 | F8 | F9 | F10 | F11 | F12
```

**Reply by ristomatti** [reply]
I didn't this these either but you might get away with just this:

```starlark
[modifier = "!Modifier"]
  ? Tap[A] >> A
  Hold[A]  >> !A Meta
```

Or similarly as my example above:

```starlark
VMeta = Virtual0

[modifier = "VMeta"]
  Alpha    >> !A Meta{Alpha}
  Digit    >> !A Meta{Digit}

[modifier = "!Modifier"]
  ? Tap[A] >> A
  Hold[A]  >> VMeta
```

**Reply by maykot** [reply]
I should have mentioned, I did try that already. Your config was very helpful, btw, I stole a few things from you :)

I thought it didn't do what I wanted, but now that I'm looking at your code I think I might have missed some stuff when I tried it. I'll give it another try in a bit and see if it works, thanks!

**Reply by maykot** [reply]
> [modifier = "!Modifier"]
>   ? Tap[A] >> A
>   Hold[A]  >> !A Meta

Yeah, I must have screwed something up when I tried this before, this is much closer to the behavior I'm looking for.

Close enough that I'll probably start using this for now. I can even combine home row mods, the only slightly annoying thing is that I have to respect the hold period for each mod. I can't press two mods at the same time and then the key I want, I have to press one, wait for the hold period, press the next, wait for the hold period again, press the final key.

But this is already a massive improvement from what I had before.
Thanks again!

**Reply by ristomatti** [reply]
What if you change `Hold[A] >> !A Meta` to `A{Any} >> !A Meta{Any}`? Does it help with the timing or break it altogether? I fiddled **a lot** with this and getting it to work with the Space to Meta binding. To me it seems the solution pretty much requires there to be a delay somewhere.

**Reply by maykot** [reply]
> What if you change Hold[A] >> !A Meta to A{Any} >> !A Meta{Any}? Does it help with the timing or break it altogether?

It definitely improves the timings, but then I lose the ability to combine home row mods.

> To me it seems the solution pretty much requires there to be a delay somewhere.

I guess that's probably true. It's just a matter of where the delay is. The reason why I mentioned kanata's [tap-hold-release](https://github.com/jtroo/kanata/blob/main/docs/config.adoc#tap-hold) is because it's functionally pretty much what I wanted ([this](https://github.com/jtroo/kanata/discussions/718#discussioncomment-8357280) config, for example, if you replace the `tap-hold` by `tap-hold-release`), I just don't know if it's possible to achieve the same with keymapper because I don't have enough experience with it yet.

**Reply by ristomatti** [reply]
I got curious and wanted to see if I could come up with a solution. Stubborn as I am, I ended up spending more time on this I'd like to admit (a lot longer than a couple of hours)... :grimacing:

I came up with two solution, each with their own shortcomings. (Caveat: I didn't thoroughly test them, so there might be even bigger issues than the ones I noticed. I didn't test them in isolation from my other config either.)

But anyhow, let's go through them.

**First attempt:**

```ini
HMeta    = A
HShift   = S
HAlt     = D
HControl = F

HMod     = A | S | D | F

Tap      = $0{!120ms}
Hold     = $0{130ms}

[default]
  HMod !60ms HMod      >> HMod HMod
  ? Tap[HMod]          >> HMod

  ? (HMeta HShift)     >> !HMod (Meta Shift)
  ? (HMeta HAlt)       >> !HMod (Meta AltLeft)
  ? (HMeta HControl)   >> !HMod (Meta Control)
  ? (HShift HAlt)      >> !HMod (Shift AltLeft)
  ? (HShift HControl)  >> !HMod (Shift Control)
  ? (HAlt HControl)    >> !HMod (AltLeft Control)

  Hold[HMeta]          >> !HMeta Meta
  Hold[HShift]         >> !HShift Shift
  Hold[HAlt]           >> !HAlt AltLeft
  Hold[HControl]       >> !HControl Control
```

This should solve the issue of having to wait through the hold period when pressing two home row mods. In fact, there's no hold time. There's a big issue with this one though: if two home row mod keys are typed fast in succession, it triggers the modifier combination, at least on my QMK mechanical keyboard thi is.

**Second attempt:**

In my testing, this monstrosity should slightly improve my initial suggestion. With this, you only need to wait the hold time for the first home row mod.

```ini
HMeta    = A
HShift   = S
HAlt     = D
HControl = F

HMod     = A | S | D | F

VMeta    = Virtual50
VShift   = Virtual51
VAlt     = Virtual52
VControl = Virtual53

Tap      = $0{!120ms}
Hold     = $0{130ms}

[modifier = "VMeta"]
  ContextActive   >> Meta
  HShift          >> Shift
  HAlt            >> AltLeft
  HControl        >> Control

[modifier = "VShift"]
  ContextActive   >> Shift
  HMeta           >> Meta
  HAlt            >> AltLeft
  HControl        >> Control

[modifier = "VAlt"]
  ContextActive   >> AltLeft
  HMeta           >> Meta
  HShift          >> Shift
  HControl        >> Control

[modifier = "VControl"]
  ContextActive   >> Control
  HMeta           >> Meta
  HShift          >> Shift
  HAlt            >> AltLeft

[default]
  ? Tap[HMod]     >> HMod

  Hold[HMeta]     >> VMeta ^ VMeta
  Hold[HShift]    >> VShift ^ VShift
  Hold[HAlt]      >> VAlt ^ VAlt
  Hold[HControl]  >> VControl ^ VControl
```

What I didn't yet mention though was an odd issue that affects both attempts. All the other two home row mod combinations seemed to work, but `HShift{HAlt}` / `S{D}` does not output `Shift+Alt`, but `Shift+Meta`. It only occurred when pressed in that order. `HAlt{HShift}` had the expected output of `Alt+Shift`. :thinking:    

~~This might be due to other mappings on my config, but my brain is melting at this point and I'm not going to test if that is the case (at least not not today).~~

**Edit:** I did end up testing the second attempt in isolation. I was still able to reproduce the behavior.

**Reply by ristomatti** [reply]
I left out my very first attempt which I thought should have worked, but it didn't. I'll throw it out anyway in case @houmain wants to shed some light why it doesn't and/or if it can be adjusted to work.


```ini
HMeta    = A
HShift   = S
HAlt     = D
HControl = F

HMod     = A | S | D | F

Tap      = $0{!120ms}
Hold     = $0{130ms}

[default]
  ? Tap[HMod]            >> HMod

  Hold[(HMeta HShift)]   >> !HMod (Meta Shift)
  Hold[(HMeta HAlt)]     >> !HMod (Meta AltLeft)
  Hold[(HMeta HControl)] >> !HMod (Meta Control)
  Hold[(HShift HAlt)]    >> !HMod (Shift AltLeft)
  Hold[(HAlt HControl)]  >> !HMod (AltLeft Control)

  Hold[HMeta]            >> !HMeta Meta
  Hold[HShift]           >> !HShift Shift
  Hold[HAlt]             >> !HAlt AltLeft
  Hold[HControl]         >> !HControl Control
```

With this, two home row mods can be pressed at the same time, avoiding the second wait period. The problem is that the last 4 mappings stop working, rendering this useless.

**Reply by houmain** [reply]
Very interesting!
I tried your configurations. 
The first did not work at all for me. It immediately starts repeating 's' when I hold `S`.
The second works really well, I could not reproduce the strange S/D rollover problem. 🤔

> I left out my very first attempt which I thought should have worked, but it didn't. I'll throw it out anyway in case @houmain wants to shed some light why it doesn't and/or if it can be adjusted to work.

I think the problem is, that the following minimal sample does not work either:
```
(F1 F2){200ms} >> A
F1{200ms} >> B
F2{200ms} >> C
```
When only `F1` or `F2` is pressed it waits indefinitely until the other is pressed.

Swapping the mappings does not work either, since it never gets to the third mapping:
```
F1{200ms} >> B
F2{200ms} >> C
(F1 F2){200ms} >> A
```

Fortunately the following configuration works as desired:
```
!F2 F1{200ms} >> B
!F1 F2{200ms} >> C
(F1 F2){200ms} >> A
```

FYI: I just released a new version, which has a `@done` directive, which allows to stop parsing the configuration. I use it for quickly trying out these short configurations directly at the top of my own configuration file.

**Reply by maykot** [reply]
> I got curious and wanted to see if I could come up with a solution. Stubborn as I am, I ended up spending more time on this I'd like to admit (a lot longer than a couple of hours)... 😬

Ah, and so I finally know how it feels to be on the other side of a nerd snipe :rofl:. Yeah, I've also spent way more time than I'd care to admit on similar projects in the past. Thank you for your time!

Unfortunately... neither of your configs worked for me. The same thing that happened to houmain happened to me for your first config. For the second, home row keys don't behave as I would expect them to if they are the last to be pressed in a home row combo, e.g. `+S +A -A` isn't mapped to `+Shift +A -A`, both keys are treated as mods instead.

> `!F2 F1{200ms} >> B
!F1 F2{200ms} >> C
(F1 F2){200ms} >> A`

I quickly tried this in combination with `? Tap[HMod] >> HMod` and it seemed to work really well. I still need to test it a bit more though. I'll hopefully be able to do it later today or maybe tomorrow.

**Reply by ristomatti** [reply]
> Ah, and so I finally know how it feels to be on the other side of a nerd snipe 🤣

The times I've gotten too carried away solving someone else's issues. Then coming back with a solution only to realise they had already lost interest or worked around it. :melting_face:  

> The first did not work at all for me. It immediately starts repeating 's' when I hold `S`.

It doesn't work for me either. I must have missed some key bit there. I originally intended to only send the second one, I dug the earlier attempt back pressing "undo" enough times for it to look familiar. :smile:

I tried the second one again to test the Meta key issue. I still see it with both `xev` and https://keyevents.netlify.app/. The funny thing is that it was not `Meta/Win/Super` but instead `Meta` as in `Alt/Meta` :thinking:. In fact, just pressing `Shift{Alt}` w/o Keymapper triggers it. I was too tried to remember `xev` shows `Meta` as `Super_L` :facepalm: 

Press/release Meta:

```sh
KeyPress event, serial 25, synthetic NO, window 0x7400001,
    state 0x0, keycode 133 (keysym 0xffeb, Super_L), same_screen YES,
KeyRelease event, serial 28, synthetic NO, window 0x7400001,
    state 0x40, keycode 133 (keysym 0xffeb, Super_L), same_screen YES,
```

Press/release Shift+Alt
```sh
KeyPress event, serial 28, synthetic NO, window 0x7400001,
    state 0x0, keycode 62 (keysym 0xffe2, Shift_R), same_screen YES,
KeyPress event, serial 28, synthetic NO, window 0x7400001,
    state 0x1, keycode 64 (keysym 0xffe7, Meta_L), same_screen YES,
KeyRelease event, serial 28, synthetic NO, window 0x7400001,
    state 0x9, keycode 64 (keysym 0xffe7, Meta_L), same_screen YES,
KeyRelease event, serial 28, synthetic NO, window 0x7400001,
    state 0x1, keycode 62 (keysym 0xffe2, Shift_R), same_screen YES,
```

Same in Keyzz web test:

![image](https://github.com/user-attachments/assets/6c7f3732-8ff8-4715-af37-5cf7af409a93)

This is the first time I see this. I wonder if it's a general Linux thing or something to do with my specific setup, e.g. the layout I use "English (intl., with AltGr dead keys)".

**Reply by ristomatti** [reply]
> Fortunately the following configuration works as desired:
> 
> ```
> !F2 F1{200ms} >> B
> !F1 F2{200ms} >> C
> (F1 F2){200ms} >> A
> ```

I couldn't get this idea to play together with the last config I sent though. I'm not going to go through the various iterations this time for there were too many. 😆

While it might not be useful as a config, I got pretty interesting result with this:


```ini
HMeta    = A
HShift   = S
HControl = D
HAlt     = F

HMod     = A | S | D | F

Tap      = $0{!120ms}
Hold     = $0{130ms}

[default]
  ? Tap[HMod]    >> HMod

[stage]
  Hold[HAlt]     >> !HMod AltLeft

[stage]
  Hold[HControl] >> !HMod Control

[stage]
  Hold[HShift]   >> !HMod Shift

[stage]
  Hold[HMeta]    >> !HMod Meta
```

Note that I swapped `D` to be `HControl` and `F` to be `HAlt`. No particular reason, expect that it felt it made more sense here. The order of the different stages is deliberate. It allows triggering the mods in a rolling manner `A{S}` -> `Meta{Shift}`, `A{S{D}}` -> `Meta{Shift{Control}}` etc. If pressed the other way around, the mods just repeat in quick succession.

The bad part is that if the modifiers are combined with some other key, there will also be keypress events for each of the home row mod letters. :rofl:

P.S. The `@done` directive is nifty!

**Reply by ristomatti** [reply]
Sorry @maykot for all these useless configs, but I have to say this thread has been great!

**Reply by maykot** [reply]
> This is the first time I see this. I wonder if it's a general Linux thing or something to do with my specific setup, e.g. the layout I use "English (intl., with AltGr dead keys)".

I'm also on Linux using the same layout and I can confirm this happens to me too. Are you on KDE too? I did a quick [search](https://www.reddit.com/r/archlinux/comments/1g3kf50/deactivate_shiftalt_functioning_as_meta_key/) and it looks like this could be a KDE thing, not sure if it's a bug or intended. Kinda crazy that this has never been an issue for me. I guess it's because only `Shift + Alt` is affected and not `Alt + Shift`.

> Sorry @maykot for all these useless configs, but I have to say this thread has been great!

No worries, nothing like a good puzzle, I know the feeling. And even though your configs didn't work they still helped me.

**Reply by maykot** [reply]
> `!F2 F1{200ms} >> B !F1 F2{200ms} >> C (F1 F2){200ms} >> A`

After playing with this for a bit I managed to come up with this behemoth of a config:
```
# Home row mods

HRMetaL       = A
HRAltL        = S
HRControlL    = D
HRShiftL      = F

# Pending implementation
HRDiaL        = G
HRDiaR        = H

HRShiftR      = J
HRControlR    = K
HRAltR        = L
HRMetaR       = Semicolon

HRModL        = HRMetaL | HRAltL | HRControlL | HRShiftL | HRDiaL
HRModR        = HRMetaR | HRAltR | HRControlR | HRShiftR | HRDiaR
HRMod         = HRModL | HRModR

HRNotMetaL    =           HRAltL | HRControlL | HRShiftL | HRDiaL | HRModR
HRNotAltL     = HRMetaL |          HRControlL | HRShiftL | HRDiaL | HRModR
HRNotControlL = HRMetaL | HRAltL |              HRShiftL | HRDiaL | HRModR
HRNotShiftL   = HRMetaL | HRAltL | HRControlL            | HRDiaL | HRModR
HRNotDiaL     = HRMetaL | HRAltL | HRControlL | HRShiftL          | HRModR

HRNotMetaR    =           HRAltR | HRControlR | HRShiftR | HRDiaR | HRModL
HRNotAltR     = HRMetaR |          HRControlR | HRShiftR | HRDiaR | HRModL
HRNotControlR = HRMetaR | HRAltR |              HRShiftR | HRDiaR | HRModL
HRNotShiftR   = HRMetaR | HRAltR | HRControlR            | HRDiaR | HRModL
HRNotDiaR     = HRMetaR | HRAltR | HRControlR | HRShiftR          | HRModL

[default]
    ? Tap[HRMod]                    >> HRMod

    !HRNotMetaL    Hold[HRMetaL]    >> !HRMetaL    MetaLeft
    !HRNotAltL     Hold[HRAltL]     >> !HRAltL     AltLeft
    !HRNotControlL Hold[HRControlL] >> !HRControlL ControlLeft
    !HRNotShiftL   Hold[HRShiftL]   >> !HRShiftL   ShiftLeft

    Hold[(HRMetaL    HRAltL)]       >> !HRMod (MetaLeft    AltLeft)
    Hold[(HRMetaL    HRControlL)]   >> !HRMod (MetaLeft    ControlLeft)
    Hold[(HRMetaL    HRShiftL)]     >> !HRMod (MetaLeft    ShiftLeft)
    Hold[(HRAltL     HRControlL)]   >> !HRMod (AltLeft     ControlLeft)
    Hold[(HRAltL     HRShiftL)]     >> !HRMod (AltLeft     ShiftLeft)
    Hold[(HRControlL HRShiftL)]     >> !HRMod (ControlLeft ShiftLeft)

    !HRNotMetaR    Hold[HRMetaR]    >> !HRMetaR    MetaRight
    !HRNotAltR     Hold[HRAltR]     >> !HRAltR     AltLeft
    !HRNotControlR Hold[HRControlR] >> !HRControlR ControlRight
    !HRNotShiftR   Hold[HRShiftR]   >> !HRShiftR   ShiftRight

    Hold[(HRMetaR    HRAltR)]       >> !HRMod (MetaRight    AltLeft)
    Hold[(HRMetaR    HRControlR)]   >> !HRMod (MetaRight    ControlRight)
    Hold[(HRMetaR    HRShiftR)]     >> !HRMod (MetaRight    ShiftRight)
    Hold[(HRAltR     HRControlR)]   >> !HRMod (AltLeft      ControlRight)
    Hold[(HRAltR     HRShiftR)]     >> !HRMod (AltLeft      ShiftRight)
    Hold[(HRControlR HRShiftR)]     >> !HRMod (ControlRight ShiftRight)
```
And this doesn't even take into account all possible key combos. There's probably an easier way to do this, but that's what worked for me.

I tested this reasonably well and it seems to be working as intended. The only caveat is something I hadn't mentioned up until now because I didn't want to complicate the problem even more and I thought it would be kinda trivial once we had a solution for the main problem. As it stands I cannot repeat a home row key by holding it down, that's just unavoidable in this setup. To address this I was planning on adding something along the lines of `HRMetaL !120ms HRMetaL >> HRMetaL HRMetaL` to the top of the default section. The intention was that double tapping a key and holding it would trigger the repeat of the original key. This doesn't seem to work though, and I also couldn't get any variation involving `?`, adding a hold or changing where the double tap remap was defined to play along nicely with my config. I think I'm probably misunderstanding how `?` and the evaluation order works.

**Reply by ristomatti** [reply]
I hate to say this but this is essentially exactly what I first tried as well and has the same issue I had as well. Depending on your shortcut usage patterns it might not be that big of an issue though. Triggering output of `Meta{Shift{1}}` works, but can you get it to output `Control{Alt{1}}` or `Alt{Shift{1}}` (using HR mods keys)? I can't.

 > As it stands I cannot repeat a home row key by holding it down, that's just unavoidable in this setup. To address this I was planning on adding something along the lines of HRMetaL !120ms HRMetaL >> HRMetaL HRMetaL to the top of the default section. The intention was that double tapping a key and holding it would trigger the repeat of the original key. 

I had the exact same thought process and encountered the same issue. In fact, I've ran across this earlier on with my own config. If you only need to do this occasionally, you can get a key to repeat if you first type some key that doesn't output or do anything in the given context, then quickly follow up with the key you want to repeat. But perhaps a better workaround is to modify your config to include something like this:

```ini
RepeatMod     = Virtual50

[modifier = "RepeatMod"]
    HRMod >> HRMod

[default]
    Hold["Tab"] >> RepeatMod ^ RepeatMod

    ? Tap[HRMod]                    >> HRMod
```

You could also try if using `[stage]` either before or after the HRM section and try if you can come up with something that works when evaluated in a different stage.

**Edit:** changed `RepeatMod` to a virtual modifier (should not edit after pasting a working snippet)

**Reply by ristomatti** [reply]
> Are you on KDE too?

Nope, Gnome with i3 WM on Ubuntu 20.04.

**Reply by ristomatti** [reply]
I also verified now with `evtest` that `Shift+Alt` is just that, `Shift` and `Alt` on the input device level. I also tried Finnish layout and still get Meta_L, so perhaps it's something happening on X11 level :shrug:.

**Reply by maykot** [reply]
> I hate to say this but this is essentially exactly what I first tried as well and has the same issue I had as well. Depending on your shortcut usage patterns it might not be that big of an issue though. Triggering output of Meta{Shift{1}} works, but can you get it to output Control{Alt{1}} or Alt{Shift{1}} (using HR mods keys)? I can't.

You're totally right. Now I'm 100% certain my config is gaslighting me. I was sure I had tested `J{K{V}}`  or `K{J{V}}` and it correctly pasted to the terminal (i.e. `Control{Shift{V}}`), but no, it definitely doesn't in either order. It's kinda funny that some combinations work though. `A{D{ArrowLeft}}` is correctly interpreted as `Meta{Control{ArrowLeft}}`, but `D{A{ArrowLeft}}` isn't. I guess it's back to the drawing board then.

> But perhaps a better workaround is to modify your config to include something like this:

Oh, that might also be a nice alternative. I'll try this once I have a working config.

**Reply by maykot** [reply]
> I also verified now with evtest that Shift+Alt is just that, Shift and Alt on the input device level. I also tried Finnish layout and still get Meta_L, so perhaps it's something happening on X11 level 🤷.

I'm on Wayland. I just checked keyzz on two different browsers, one running directly on Wayland and the other running under Xwayland, the issue only happened on the browser running under Xwayland. So yeah, it seems it's an X11 issue.

**Comment by houmain** [comment]
Does `A{E !E} >> Meta{E}` behave as you intend?
I just tried `A{Any !Any} >> Meta{Any}` but that does not work yet.

**Reply by maykot** [reply]
> Does A{E !E} >> Meta{E} behave as you intend?

Almost perfectly. The only thing that doesn't seem to work the way I intended is using this in combination with other modifiers. My intention was to be able to use this in combination with other, possibly home row, modifiers as well. The first non-mod key to be released would be the trigger for the binding.
To use some concrete examples: I'd be able to `+A +Shift -Shift +Control +E -E` and it would trigger `Meta{Control{E}}` (`+` and `-` are press and release).  With the example you gave `+A +Shift` triggers `A` immediately.

I realize this might be asking a bit too much, so no worries if it's not possible.

> I just tried A{Any !Any} >> Meta{Any} but that does not work yet.

Yeah, that was one of the first things I tried.


Thank you for this project, btw, it's amazing! The device filters have been specially useful for me.

**Reply by ristomatti** [reply]
> . To use some concrete examples: I'd be able to `+A +Shift -Shift +Control +E -E` and it would trigger `Meta{Control{E}}` (`+` and `-` are press and release). With the example you gave `+A +Shift` triggers `A` immediately.

Would your ideal setup also support holding down multiple Home Row Mods? How would e.g. `Meta{F}` be handled when `A{F}` if we assume `A` to be `Meta` and `F` to be another mod key?

**Reply by maykot** [reply]
> Would your ideal setup also support holding down multiple Home Row Mods?

Yes, definitely.

> How would e.g. Meta{F} be handled when A{F} if we assume A to be Meta and F to be another mod key?

This is a good example. I don't know how to implement it in keymapper, but in terms of key press and release the logic would be something like this:
When a home row mod key is pressed defer its evaluation until a key is pressed and then released with no other events in between. The key that was pressed and then released retains its default value, the rest are treated as mods. In the example above there are a few ways `A` and `F` could be stringed together, `+A +F -F` >> `+Meta +F -F`, `+F +A -A` >> `+Control +A -A` (assuming `F` is `Control`), `+A +F -A` >> `+A +F -A`.
This is probably missing a bunch of edge cases that I'm very likely overlooking, but the rough idea is there.

**Reply by MatthiasGrandl** [reply]
@maykot have you found a solution for combining multiple home row mods? was super happy initially after migrating my config from kmonad to keymapper until I stumbled across this limitation lol

**Reply by maykot** [reply]
No, I haven't. But I haven't tried ristomatti's [latest suggestion](https://github.com/houmain/keymapper/discussions/187#discussioncomment-11706611) yet.

**Reply by MatthiasGrandl** [reply]
I guess it probably does work, just wanted to avoid defining all combinations separately haha.

**Reply by MatthiasGrandl** [reply, marked as answer]
PS: I just did and it seems to be working pretty well. If anyone has suggestions for improvements, lemme know.

```

# Homerow Mods
FakeAlt = S | L
FakeControl = D | K
FakeShift = F | J
FakeMeta = A | Semicolon

FakeMeta{Any !Any} >> Meta{Any}
FakeAlt{Any !Any} >> Alt{Any}
FakeControl{Any !Any} >> Control{Any}
FakeShift{Any !Any} >> Shift{Any}

# All combinations of 4
(FakeMeta FakeAlt FakeControl FakeShift){Any !Any} >> (Meta Alt Control Shift){Any}

# All combinations of 3
(FakeMeta FakeAlt FakeControl){Any !Any} >> (Meta Alt Control){Any}
(FakeMeta FakeAlt FakeShift){Any !Any} >> (Meta Alt Shift){Any}
(FakeMeta FakeControl FakeShift){Any !Any} >> (Meta Control Shift){Any}
(FakeAlt FakeControl FakeShift){Any !Any} >> (Alt Control Shift){Any}

# All combinations of 2
(FakeMeta FakeAlt){Any !Any} >> (Meta Alt){Any}
(FakeMeta FakeControl){Any !Any} >> (Meta Control){Any}
(FakeMeta FakeShift){Any !Any} >> (Meta Shift){Any}
(FakeAlt FakeControl){Any !Any} >> (Alt Control){Any}
(FakeAlt FakeShift){Any !Any} >> (Alt Shift){Any}
(FakeControl FakeShift){Any !Any} >> (Control Shift){Any}
```

**Reply by houmain** [reply]
I think this question has come up more than once and this looks like a clean solution to me. I am thinking about putting this configuration and other common patterns in the Wiki.

**Reply by maykot** [reply]
```
# Homerow Mods
FakeAlt = S | L
FakeControl = D | K
FakeShift = F | J
FakeMeta = A | Semicolon
 ...
```
Can confirm, woks like a charm. Thank you, @MatthiasGrandl!

**Reply by MatthiasGrandl** [reply]
You are welcome! Thanks for the amazing software @houmain !

**Reply by Zireael07** [reply]
@houmain A wiki with samples (including samples on how to use keymapperctl) would be awesome!

**Reply by ristomatti** [reply]
@Zireael07 Added one example use case to wiki https://github.com/houmain/keymapper/wiki#text-expansion.

**Reply by ristomatti** [reply]
Also added CapsWord modifier: https://github.com/houmain/keymapper/wiki#capsword. @houmain feel free to remove if wished. I didn't realize the changes save without triggering a review process. 🙂

**Reply by ristomatti** [reply]
AutoShift: https://github.com/houmain/keymapper/wiki#autoshift
Key group aliases: https://github.com/houmain/keymapper/wiki#key-group-aliases

**Reply by ristomatti** [reply]
Vim-style arrows: https://github.com/houmain/keymapper/wiki#vim-style-arrows

**Reply by ristomatti** [reply]
Vim-motions (modal variant): https://github.com/houmain/keymapper/wiki#vim-motions-modal

**Reply by houmain** [reply]
> feel free to remove if wished.

No, looks great!

>feel free to remove if wished. I didn't realize the changes save without triggering a review process.

Currently anyone can contribute. Hopefully only improvements and not too much work to clean up. We could restrict editing to collaborators any time.

**Comment by ristomatti** [comment]
@maykot If you're still on this quest, you might want to take a look at https://github.com/houmain/keymapper/issues/210#issuecomment-2566001947 in case that config would help in solving this. (I didn't try it myself though)

**Reply by maykot** [reply]
Oh, sorry, I missed this. My New Year was a bit hectic. I'll give it a try when I have sometime to play around with my config again.
